src/rtc.rs

```diff
@@ -0,0 +1,458 @@
+        let (mt, mu) = bcd2_encode(date.month())?;
+        let (dt, du) = bcd2_encode(date.day())?;
+
+        self.modify(|regs| {
+            regs.dr.write(|w| unsafe {
+                w.dt().bits(dt);
+                w.du().bits(du);
+                w.mt().bit(mt > 0);
+                w.mu().bits(mu);
+                w.yt().bits(yt);
+                w.yu().bits(yu)
+            })
+        });
+
+        Ok(())
+    }
+
+    fn set_datetime(&mut self, date: &NaiveDateTime) -> Result<(), Self::Error> {
+        if date.year() < 1970 {
+            return Err(Error::InvalidInputData);
+        }
+
+        self.set_24h_fmt();
+        let (yt, yu) = bcd2_encode((date.year() - 1970) as u32)?;
+        let (mt, mu) = bcd2_encode(date.month())?;
+        let (dt, du) = bcd2_encode(date.day())?;
+
+        let (ht, hu) = bcd2_encode(date.hour())?;
+        let (mnt, mnu) = bcd2_encode(date.minute())?;
+        let (st, su) = bcd2_encode(date.second())?;
+
+        self.modify(|regs| {
+            regs.dr.write(|w| unsafe {
+                w.dt().bits(dt);
+                w.du().bits(du);
+                w.mt().bit(mt > 0);
+                w.mu().bits(mu);
+                w.yt().bits(yt);
+                w.yu().bits(yu)
+            });
+            regs.tr.write(|w| unsafe {
+                w.ht().bits(ht);
+                w.hu().bits(hu);
+                w.mnt().bits(mnt);
+                w.mnu().bits(mnu);
+                w.st().bits(st);
+                w.su().bits(su);
+                w.pm().clear_bit()
+            })
+        });
+
+        Ok(())
+    }
+
+    fn get_seconds(&mut self) -> Result<u8, Self::Error> {
+        let tr = self.regs.tr.read();
+        let seconds = bcd2_decode(tr.st().bits(), tr.su().bits());
+        Ok(seconds as u8)
+    }
+
+    fn get_minutes(&mut self) -> Result<u8, Self::Error> {
+        let tr = self.regs.tr.read();
+        let minutes = bcd2_decode(tr.mnt().bits(), tr.mnu().bits());
+        Ok(minutes as u8)
+    }
+
+    fn get_hours(&mut self) -> Result<Hours, Self::Error> {
+        let tr = self.regs.tr.read();
+        let hours = bcd2_decode(tr.ht().bits(), tr.hu().bits());
+        if self.is_24h_fmt() {
+            return Ok(Hours::H24(hours as u8));
+        }
+        if !tr.pm().bit() {
+            return Ok(Hours::AM(hours as u8));
+        }
+        Ok(Hours::PM(hours as u8))
+    }
+
+    fn get_time(&mut self) -> Result<NaiveTime, Self::Error> {
+        self.set_24h_fmt();
+        let seconds = self.get_seconds().unwrap();
+        let minutes = self.get_minutes().unwrap();
+        let hours = hours_to_u8(self.get_hours()?)?;
+
+        Ok(NaiveTime::from_hms(
+            hours.into(),
+            minutes.into(),
+            seconds.into(),
+        ))
+    }
+
+    fn get_weekday(&mut self) -> Result<u8, Self::Error> {
+        let dr = self.regs.dr.read();
+        let weekday = bcd2_decode(dr.wdu().bits(), 0x00);
```

Comment:
Shouldn't the order of the arguments be inverse here?
```suggestion
        let weekday = bcd2_decode(0x00, dr.wdu().bits());
```
or maybe even better just:
```suggestion
        let weekday = dr.wdu().bits();
```

Reply:
Reviews are definitely useful - I don't think I used the day-of-week functionality at all in my code, and I may have skipped reviewing that it worked correctly. I'll check this out in the next few days.

Reply:
Will you please also check the following: in section 17.3.6 Reading the calendar, there is:
```
To ensure consistency between the 3 values, reading either RTC_SSR or RTC_TR locks
the values in the higher-order calendar shadow registers until RTC_DR is read.
```
I noticed this behavior in my own implementation of RTC, so I had to make sure I read DR after I read SSR or TR, but I can't see this provision in you implementation. So e.g. if I would read only time for several days in a row then a DR read will return a date from when TR was first read and not the current date.

Reply:
Yes, this is why I reversed the order in get_datetime from date, then time, to time, then date. Ugh, that really potentially muddies the code quite a bit.

Reply:
Well, you can just make sure to just read (and discard) DR in every time-only function after the TR is read.